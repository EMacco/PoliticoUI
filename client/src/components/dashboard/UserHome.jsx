import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import AuthHeader from '@components/layouts/AuthHeader';
import AuthFooter from '@components/layouts/AuthFooter';
import AuthNav from '@components/layouts/AuthNav';
import Card from '../layouts/Card';
import { fetchInterestedOffice, fetchPageVotes } from '@actions/dashboard';

class UserHome extends Component {
  componentDidMount() {
    const { user } = this.props.auth;
    this.props.fetchInterestedOffice({ user });
    this.props.fetchPageVotes(user);
  }

  render() {
    const { votes, interest } = this.props;
    const { officeId, officeName, image, numberOfCandidates, electionDate } = interest;

    const interestedOfficeContent = [
      {
        officeId,
        officeName,
        numberOfCandidates,
        type: 'link'
      },
      {
        title: 'Date',
        value: electionDate
      }
    ];

    let interestCard = '';

    if (officeId !== undefined) {
      interestCard = (
        <Fragment>
          <Card name={officeName} image={image} content={interestedOfficeContent} />
          <Card hidden="true" />
        </Fragment>
      );
    }

    let votesCards = [];
    this.props.votes.map((vote, ind) => {
      const { partyLogo, name, partyName, officeName, result, resultStyle, resultString } = vote;

      const voteContent = [
        {
          title: 'Party',
          value: partyName
        },
        {
          title: 'Office',
          value: officeName
        },
        {
          title: 'Result',
          value: result
        },
        {
          text: resultString,
          style: `${resultStyle}-election`,
          type: 'span'
        }
      ];

      votesCards.push(
        <Fragment key={Math.random().toString()}>
          <Card name={name} image={partyLogo} content={voteContent} />
          {votes.length % 2 !== 0 && ind === votes.length - 1 ? <Card hidden="true" /> : ''}
        </Fragment>
      );
    });

    return (
      <div>
        <AuthHeader />
        <main id="user-home-main">
          <div className="container">
            <AuthNav />

            <section id="user-home-content-div">
              <div id="loadingIconContainer" />
              <div className="profile-page-section">
                <h3 id="interested-office-header">Interested Office</h3>
                <div className="content-group" id="interestedOfficeSlot">
                  {interestCard}
                </div>
              </div>

              <div className="profile-page-section">
                <h3>My Votes ({votes.length})</h3>
                <div className="content-group" id="userVotesSlot">
                  {votesCards}
                </div>
              </div>
            </section>
          </div>
          <AuthFooter />
        </main>
      </div>
    );
  }
}

UserHome.propTypes = {
  fetchInterestedOffice: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  interest: PropTypes.object.isRequired,
  fetchPageVotes: PropTypes.func.isRequired,
  votes: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  interest: state.dashboard.interests,
  votes: state.dashboard.votes
});

export default connect(
  mapStateToProps,
  { fetchInterestedOffice, fetchPageVotes }
)(UserHome);
