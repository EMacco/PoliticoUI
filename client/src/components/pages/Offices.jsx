import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import AuthHeader from '../layouts/AuthHeader';
import AuthFooter from '../layouts/AuthFooter';
import AuthNav from '../layouts/AuthNav';
import Card from '../layouts/Card';
import { fetchGovernmentOffices, expressInterest } from '@actions/office';

class Offices extends Component {
  componentDidMount() {
    const {
      fetchGovernmentOffices,
      user: { id }
    } = this.props;
    fetchGovernmentOffices(id);
  }

  expressInterest = e => {
    const { expressInterest } = this.props;
    const { officeId, partyId } = JSON.parse(e.target.id);
    if (!partyId) return toast.error('Please join a political party before contesting');
    expressInterest(partyId, officeId);
  };

  render() {
    const { offices } = this.props;
    const {
      user: { partyid: partyId }
    } = this.props;

    let officesCard = [];
    offices.map((office, ind) => {
      const {
        officeId,
        officeLogo,
        officeName,
        officeType,
        electionDate,
        count,
        candidateStatus,
        officeBtn,
        alreadyApplied
      } = office;

      const officeContent = [
        {
          title: 'Type',
          value: officeType
        },
        {
          title: 'Date',
          value: electionDate
        },
        {
          officeId,
          officeName,
          numberOfCandidates: count,
          type: 'link'
        },
        {
          text: candidateStatus,
          style: officeBtn,
          type:
            candidateStatus === 'Approved candidate' || candidateStatus === 'Awaiting approval'
              ? 'span'
              : !alreadyApplied
              ? 'button'
              : 'empty',
          onClick: !alreadyApplied ? this.expressInterest : null,
          id: JSON.stringify({ officeId, partyId })
        }
      ];

      officesCard.push(
        <Fragment key={Math.random().toString()}>
          <Card name={officeName} image={officeLogo} content={officeContent} />
          {offices.length % 2 !== 0 && ind === offices.length - 1 ? <Card hidden="true" /> : ''}
        </Fragment>
      );
    });

    return (
      <div>
        <AuthHeader />
        <main id="user-home-main">
          <div className="container">
            <AuthNav current="offices" />
            <div id="loadingIconContainer" />
            <section id="user-home-content-div">
              <div className="profile-page-section">
                <h3 id="interested-office-header">Government Offices ({officesCard.length})</h3>
                <div className="content-group" id="governmentOfficeSlot">
                  {officesCard}
                </div>
              </div>
            </section>
          </div>
          <AuthFooter />
        </main>
      </div>
    );
  }
}

Offices.propTypes = {
  fetchGovernmentOffices: PropTypes.func.isRequired,
  offices: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  user: state.auth.user,
  offices: state.offices.offices
});

export default connect(
  mapStateToProps,
  { fetchGovernmentOffices, expressInterest }
)(Offices);
