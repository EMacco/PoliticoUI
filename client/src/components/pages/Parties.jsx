import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import AuthHeader from '../layouts/AuthHeader';
import AuthFooter from '../layouts/AuthFooter';
import AuthNav from '../layouts/AuthNav';
import Card from '../layouts/Card';
import { fetchPoliticalParties, changeUserParty } from '@actions/party';

class Parties extends Component {
  componentDidMount() {
    const { fetchPoliticalParties } = this.props;
    fetchPoliticalParties();
  }

  joinParty = e => {
    const { changeUserParty } = this.props;
    changeUserParty(e.target.id);
  };

  render() {
    const {
      parties,
      user: { partyid }
    } = this.props;

    let partiesCard = [];
    parties.map((party, ind) => {
      const { partyId, partyLogo, partyName, hqAddress } = party;

      const member = partyid === partyId ? true : false;
      let partyBtn = 'add-party-btn';
      let partyText = 'Join Party';
      // Check if this user is a member of this party
      if (partyid === partyId) {
        partyBtn = 'casted-vote';
        partyText = 'Member';
      }

      const partyContent = [
        {
          title: 'HQ Address',
          value: hqAddress
        },
        {
          text: partyText,
          style: partyBtn,
          type: member ? 'span' : 'button',
          onClick: !member ? this.joinParty : null,
          id: partyId
        }
      ];

      partiesCard.push(
        <Fragment key={Math.random().toString()}>
          <Card name={partyName} image={partyLogo} content={partyContent} />
          {parties.length % 2 !== 0 && ind === parties.length - 1 ? <Card hidden="true" /> : ''}
        </Fragment>
      );
    });

    return (
      <div>
        <AuthHeader />
        <main id="user-home-main">
          <div className="container">
            <AuthNav current="parties" />
            <div id="loadingIconContainer" />
            <section id="user-home-content-div">
              <div className="profile-page-section">
                <h3>Political Parties ({parties.length})</h3>
                <div className="content-group" id="userVotesSlot">
                  {partiesCard}
                </div>
              </div>
            </section>
          </div>
          <AuthFooter />
        </main>
      </div>
    );
  }
}

Parties.propTypes = {
  parties: PropTypes.array.isRequired,
  fetchPoliticalParties: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  parties: state.parties.parties,
  user: state.auth.user
});

export default connect(
  mapStateToProps,
  { fetchPoliticalParties, changeUserParty }
)(Parties);
