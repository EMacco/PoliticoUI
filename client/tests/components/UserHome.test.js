import React from 'react';
import { UserHome } from '@components/pages/UserHome';

const props = {
  fetchInterestedOffice: () => {},
  auth: {},
  interest: {
    officeId: 1,
    officeName: 'Office',
    image: '',
    numberOfCandidates: 4,
    electionDate: '5th May'
  },
  fetchPageVotes: () => {},
  votes: [
    {
      partyLogo: '',
      name: 'John',
      partyName: 'Party',
      officeName: 'Office',
      result: 5,
      resultStyle: 'won',
      resultString: 'won'
    }
  ],
  user: { partyid: 2 }
};

const setup = () => {
  const wrapper = shallow(<UserHome {...props} />);
  return wrapper;
};

describe('<UserHome />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });

  it('should render the UserHome component correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('UserHome')).toBeTruthy();
  });
});
