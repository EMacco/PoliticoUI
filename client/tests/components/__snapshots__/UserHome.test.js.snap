// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserHome /> should render the UserHome component correctly 1`] = `
<div>
  <Connect(withRouter(AuthHeader)) />
  <main
    id="user-home-main"
  >
    <div
      className="container"
    >
      <Connect(withRouter(AuthNav))
        current="profile"
      />
      <section
        id="user-home-content-div"
      >
        <div
          id="loadingIconContainer"
        />
        <div
          className="profile-page-section"
        >
          <h3
            id="interested-office-header"
          >
            Interested Office
          </h3>
          <div
            className="content-group"
            id="interestedOfficeSlot"
          >
            <Card
              content={
                Array [
                  Object {
                    "numberOfCandidates": 4,
                    "officeId": 1,
                    "officeName": "Office",
                    "type": "link",
                  },
                  Object {
                    "title": "Date",
                    "value": "5th May",
                  },
                ]
              }
              hidden="false"
              image=""
              name="Office"
            />
            <Card
              content={Array []}
              hidden="true"
              image=""
              name=""
            />
          </div>
        </div>
        <div
          className="profile-page-section"
        >
          <h3>
            My Votes (
            1
            )
          </h3>
          <div
            className="content-group"
            id="userVotesSlot"
          >
            <Card
              content={
                Array [
                  Object {
                    "title": "Party",
                    "value": "Party",
                  },
                  Object {
                    "title": "Office",
                    "value": "Office",
                  },
                  Object {
                    "title": "Result",
                    "value": 5,
                  },
                  Object {
                    "style": "won-election",
                    "text": "won",
                    "type": "span",
                  },
                ]
              }
              hidden="false"
              image=""
              name="John"
            />
            <Card
              content={Array []}
              hidden="true"
              image=""
              name=""
            />
          </div>
        </div>
      </section>
    </div>
    <AuthFooter />
  </main>
</div>
`;
